### PR TITLE
fix: search local contacts

### PR DESCRIPTION
**Pull request by beyond009**
## Description

https://mask.atlassian.net/browse/MF-5431
Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/extension/popups/pages/Friends/Home/index.tsx [hunk 1]
```diff
@@ -1,21 +1,41 @@
...
+
+        return Fuse.create(records, {
+            keys: ['profile.userId'],
+            isCaseSensitive: false,
+            ignoreLocation: true,
+            threshold: 0,
+        })
+            .search(keyword)
+            .map((item) => item.item)
+    }, [keyword, records, type])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Friends/Home/index.tsx]: Separate `Fuse.create` and `fuse.searh`.

File: packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx [hunk 2]
```diff
@@ -1,43 +1,82 @@
...
 
-export type NextIDPersonaBindingsWithIdentifier = NextIDPersonaBindings & { linkedPersona: ECKeyIdentifier } & {
+export type NextIDPersonaBindingsWithIdentifier = Omit<NextIDPersonaBindings, 'proofs'> & { proofs: Profile[] } & {
+    linkedPersona: ECKeyIdentifier
+} & {
     isLocal?: boolean
 }
 
 export function useFriendsFromSearch(
+    localSearchedResult?: Friend[],
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx]: I see you pass a NonNullable array to this hook. Let TypeScript help you to guard it.

```suggestion
    localSearchedResult: Friend[],
```

File: packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx [hunk 3]
```diff
@@ -1,43 +1,82 @@
...
+                                name: profile?.userId,
+                                created_at: '',
+                            },
+                        ],
+                        linkedPersona: item.persona,
+                        activated_at: '',
+                        persona: item.persona.publicKeyAsHex,
+                        isLocal: true,
+                    }
+                }) ?? EMPTY_LIST
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx]: You cast `item.profile` as `ProfileIdentifier`, so make sure it is on runtime.
```suggestion
        const localProfiles: NextIDPersonaBindingsWithIdentifier[] =
            localSearchedResult
                .filter((x) => x.persona.publicKeyAsHex !== currentIdentifier?.identifier.publicKeyAsHex && x.profile)
                .map((item) => {
                    const profile = item.profile as ProfileIdentifier
                    return {
                        proofs: [
                            {
                                platform: profile.network as
                                    | EnhanceableSite.Twitter
                                    | EnhanceableSite.Facebook
                                    | EnhanceableSite.Instagram,
                                identity: profile.userId,
                                is_valid: true,
                                last_checked_at: '',
                                name: profile.userId,
                                created_at: '',
                            },
                        ],
                        linkedPersona: item.persona,
                        activated_at: '',
                        persona: item.persona.publicKeyAsHex,
                        isLocal: true,
                    }
                })
```

File: packages/mask/src/extension/popups/hooks/useFriends.ts [hunk 4]
```diff
@@ -93,5 +98,32 @@ export function useFriendsPaged() {
...
 }
+
+export function useFriendFromList(searchedRecords: RelationRecord[]) {
+    const currentPersona = useCurrentPersona()
+    return useQuery(['search-local', searchedRecords], async () => {
+        const friends: Friend[] = (
+            await Promise.all(
+                searchedRecords.map(async (x) => {
+                    if (isProfileIdentifier(x.profile)) {
+                        const res = first(await Services.Identity.queryProfilesInformation([x.profile]))
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/hooks/useFriends.ts]: We could add a new query method for this specific task to improve efficiency.

File: packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx [hunk 5]
```diff
@@ -1,43 +1,82 @@
...
+                                platform: profile?.network as
+                                    | EnhanceableSite.Twitter
+                                    | EnhanceableSite.Facebook
+                                    | EnhanceableSite.Instagram,
+                                identity: profile?.userId,
+                                is_valid: true,
+                                last_checked_at: '',
+                                name: profile?.userId,
+                                created_at: '',
+                            },
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx]: Make it cleaner, this change would be also beneficial for TS performance.
You can hoist `FriendNetwork` to upper scope.

```suggestion
                .map((item) => {
                    type FriendNetwork = EnhanceableSite.Twitter
                                    | EnhanceableSite.Facebook
                                    | EnhanceableSite.Instagram
                    const profile = item.profile as ProfileIdentifier
                    return {
                        proofs: [
                            {
                                platform: profile.network as FriendNetwork,
                                identity: profile.userId,
                                is_valid: true,
                                last_checked_at: '',
                                name: profile.userId,
                                created_at: '',
                            },
```

File: packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx [hunk 6]
```diff
@@ -1,43 +1,82 @@
...
+                          activated_at: item.activated_at,
+                          persona: item.persona,
+                          isLocal: localList
+                              ? localList.some((x) => x.persona.publicKeyAsHex === identifier.publicKeyAsHex)
+                              : false,
+                      }
+                  })
+            : EMPTY_LIST
+        return uniqBy(
+            localProfiles ? localProfiles?.concat(profiles) : profiles,
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hooks/useFriendsFromSearch.tsx]:
```suggestion
            localProfiles ? localProfiles.concat(profiles) : profiles,
```

File: packages/mask/src/extension/popups/hooks/useFriends.ts [hunk 7]
```diff
@@ -93,5 +98,32 @@ export function useFriendsPaged() {
...
+                        if (res?.linkedPersona !== undefined && res?.linkedPersona !== currentPersona?.identifier)
```
Comment (UncleBill) [inline packages/mask/src/extension/popups/hooks/useFriends.ts]: These two object might not equal, comparing by public key instead.
```suggestion
                        if (res?.linkedPersona !== undefined && res?.linkedPersona.publicKeyAsHex !== currentPersona?.identifier.publicKeyAsHex)
```

**Pull request merged**